File: CraftDay.ToDo.CSharpRop/Validators/NonEmptyString.cs
using System;
using CraftDay.ToDo.CSharp.Errors;
using LanguageExt;
using LanguageExt.Common;

namespace CraftDay.ToDo.CSharpRop.Validators
{
  public class NonEmptyString
  {
    private readonly string _str;
    
    private NonEmptyString(string str)
    {
      _str = str;
    }

    public string Value() => _str;

    public static Either<Exception, NonEmptyString> From(string str)
    {
      if (!string.IsNullOrEmpty(str))
      {
        return new NonEmptyString(str);
      }
      return new ValidationException("String cannot be empty");
    }
  }
}

File: CraftDay.ToDo.FSharpRop/Validators/NonEmptyString.fs
namespace CraftDay.ToDo.FSharpRop.Validators

open CraftDay.ToDo.CSharp.Errors


type NonEmptyString = private NonEmptyString of string

module NonEmptyString =
  let create (str: string): Result<NonEmptyString, ValidationException> =
    if System.String.IsNullOrEmpty str
    then "String is not allowed to be null" |> ValidationException |> Error
    else NonEmptyString str |> Ok
  
  let value (NonEmptyString str): string =
    str